src/ssl.ml

```diff
@@ -265,91 +263,152 @@ external set_host : socket -> string -> unit = "ocaml_ssl_set1_host"
-  socket -> bigarray -> int -> int -> int = "ocaml_ssl_read_into_bigarray_blocking"
-
-external accept : socket -> unit = "ocaml_ssl_accept"
-
-external flush : socket -> unit = "ocaml_ssl_flush"
-
-external ssl_shutdown : socket -> bool = "ocaml_ssl_shutdown"
-
-let open_connection_with_context context sockaddr =
-  let domain = Unix.domain_of_sockaddr sockaddr in
-  let sock =
-    Unix.socket domain Unix.SOCK_STREAM 0 in
+(* Here is the signature of the base communication functions that are
+   implemented below in two versions *)
+module type SslBase = sig
+  val connect : socket -> unit
+  val accept : socket -> unit
+  val ssl_shutdown : socket -> bool
+  val flush : socket -> unit
+  val read : socket -> Bytes.t -> int -> int -> int
+  val read_into_bigarray : socket -> bigarray -> int -> int -> int
+  val write : socket -> Bytes.t -> int -> int -> int
+  val write_substring : socket -> string -> int -> int -> int
+  val write_bigarray : socket -> bigarray -> int -> int -> int
+end
+
+(* We now provide the implementation base communication function that release
+   the ocaml runtime lock, allowed multi-thread with blocking IO *)
+module SslReleaseBase = struct
+  external connect : socket -> unit = "ocaml_ssl_connect"
+
+  external accept : socket -> unit = "ocaml_ssl_accept"
+
+  external write : socket -> Bytes.t -> int -> int -> int = "ocaml_ssl_write"
+  external write_substring : socket -> string -> int -> int -> int
+    = "ocaml_ssl_write"
+  external write_bigarray : socket -> bigarray -> int -> int -> int
+    = "ocaml_ssl_write_bigarray"
+
+  external read : socket -> Bytes.t -> int -> int -> int = "ocaml_ssl_read"
+  external read_into_bigarray :
+    socket -> bigarray -> int -> int -> int = "ocaml_ssl_read_into_bigarray"
+
+  external flush : socket -> unit = "ocaml_ssl_flush"
+
+  external ssl_shutdown : socket -> bool = "ocaml_ssl_shutdown"
+end
+
+(* Same as above, but not releasing the lock: for non-blocking IO or
+   mono-threaded program *)
+module SslNoReleaseBase = struct
+  external connect : socket -> unit = "ocaml_ssl_connect"
+
+  external accept : socket -> unit = "ocaml_ssl_accept_blocking"
+
+  external write : socket -> Bytes.t -> int -> int -> int
+    = "ocaml_ssl_write_blocking"
+  external write_substring : socket -> string -> int -> int -> int
+    = "ocaml_ssl_write_blocking"
+  external write_bigarray : socket -> bigarray -> int -> int -> int
+    = "ocaml_ssl_write_bigarray_blocking"
+
+  external read : socket -> Bytes.t -> int -> int -> int
+    = "ocaml_ssl_read_blocking"
+
+  external read_into_bigarray :
+    socket -> bigarray -> int -> int -> int = "ocaml_ssl_read_into_bigarray_blocking"
+
+  external flush : socket -> unit = "ocaml_ssl_flush_blocking"
+
+  external ssl_shutdown : socket -> bool = "ocaml_ssl_shutdown_blocking"
+
+end
+
+(* The functor implementing communication functions from a structure of type
+   SslComBase *)
+module SslCom(C:SslBase) = struct
```

Comment:
this should be called `Make` or `Ssl_make`. `SslCom` is a bit unclear.

Reply:
> this should be called `Make` or `Ssl_make`. `SslCom` is a bit unclear.

Adressed by [28ad4b5](https://github.com/savonet/ocaml-ssl/pull/106/commits/28ad4b5b6aaaca38093c37404ab95fc4cfae4ca6). I chose SslMake ;-)